### PR TITLE
8312749: Generational ZGC: Tests crash with assert(index == 0 || is_power_of_2(index))

### PR DESCRIPTION
**Pull request by robcasloz**
This changeset ensures that the array copy stub underlying the intrinsic implementation of `Object.clone` only copies its (double-word aligned) payload, excluding the remaining object alignment padding words, when a non-default `ObjectAlignmentInBytes` value is used. This prevents the specialized ZGC stubs for `Object[]` array copy from processing undefined object alignment padding words as valid object pointers. For further details about the specific failure, see [initial analysis](https://bugs.openjdk.org/browse/JDK-8312749?focusedCommentId=14600658&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14600658) by Erik Österlund and Stefan Karlsson and comments in the regression test included in this changeset.

As a side-benefit, the changeset simplifies the array size computation logic in `GraphKit::new_array()` by decoupling computation of header size and alignment padding size.

##### Functionality

- tier1-3 (windows-x64, linux-x64, linux-aarch64, macosx-x64, and macosx-aarch64)
- tier4-5 (windows-x64, linux-x64, linux-aarch64, macosx-x64, and macosx-aarch64; ZGC-specific tests only)
- tier6-9 (linux-x64; ZGC-specific tests only)
- tier1-3, and a few custom examples, applying [JDK-8139457](https://github.com/openjdk/jdk/pull/11044) (under review) on top of this changeset

##### Performance

Tested performance on the following set of OpenJDK micro-benchmarks, on linux-x64 (for both G1 and ZGC, using different ObjectAlignmentInBytes values):

- `openjdk.bench.java.lang.ArrayClone.byteClone`
- `openjdk.bench.java.lang.ArrayClone.intClone`
- `openjdk.bench.java.lang.ArrayFiddle.simple_clone`
- `openjdk.bench.java.lang.Clone.cloneLarge`
- `openjdk.bench.java.lang.Clone.cloneThreeDifferent`

No significant regression was observed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8312749](https://bugs.openjdk.org/browse/JDK-8312749): Generational ZGC: Tests crash with assert(index == 0 || is_power_of_2(index)) (**Bug** - P3)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [5c56a5e5](https://git.openjdk.org/jdk/pull/15288/files/5c56a5e50a5cd3d15e445596639e3bf9d549eaa3)
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Contributors
 * Stefan Karlsson `<stefank@openjdk.org>`
 * Erik Österlund `<eosterlund@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15288/head:pull/15288` \
`$ git checkout pull/15288`

Update a local copy of the PR: \
`$ git checkout pull/15288` \
`$ git pull https://git.openjdk.org/jdk.git pull/15288/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15288`

View PR using the GUI difftool: \
`$ git pr show -t 15288`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15288.diff">https://git.openjdk.org/jdk/pull/15288.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15288#issuecomment-1682222892)

### CODE REVIEW

File: src/hotspot/share/gc/shared/c2/barrierSetC2.cpp [hunk 1]
```diff
@@ -675,6 +675,13 @@ void BarrierSetC2::clone(GraphKit* kit, Node* src_base, Node* dst_base, Node* si
...
   payload_size = kit->gvn().transform(new SubXNode(payload_size, offset));
+  if (is_array) {
+    // Ensure the array payload size is rounded up to the next BytesPerLong
+    // multiple when converting to double-words. This is necessary because array
+    // size does not include object alignment padding, so it might not be a
+    // multiple of BytesPerLong for sub-long element types.
+    payload_size = kit->gvn().transform(new AddXNode(payload_size, kit->MakeConX(BytesPerLong - 1)));
+  }
   payload_size = kit->gvn().transform(new URShiftXNode(payload_size, kit->intcon(LogBytesPerLong)));
   ArrayCopyNode* ac = ArrayCopyNode::make(kit, false, src_base, offset,  dst_base, offset, payload_size, true, false);
```
Comment (TobiHartmann) [inline src/hotspot/share/gc/shared/c2/barrierSetC2.cpp:686]:
```suggestion
  ArrayCopyNode* ac = ArrayCopyNode::make(kit, false, src_base, offset, dst_base, offset, payload_size, true, false);
```

**Review by TobiHartmann — APPROVED** [review]
Looks good to me!

**Review by albertnetymk — COMMENTED** [review]
If I understand it correctly, much of the diff is to ensure that `ArrayCopyNode::make` (in `BarrierSetC2::clone`) gets the correct value for the `length` arg, calculated as `align_up(array-length * elem-size, word-size) / word-size`.

I wonder if it's possible to pass the actual array length (#slots) as `length` and move the merge-bytes-to-words-copying optimization to a lower level, e.g. inside `conjoint_jbytes`. Ofc, `BarrierSetC2::clone_at_expansion` and its derived siblings need to be adjusted accordingly, e.g. to use the actual elem-type.

(Preexisting: having `ArrayCopyNode` to cover both array and instance cloning hinders the readability, IMO.)

**Comment by robcasloz** [comment]
Thanks for reviewing, Tobias!

**Review by vnkozlov — APPROVED** [review]
Looks good.

**Comment by robcasloz** [comment]
> Looks good.

Thanks for reviewing, Vladimir!

**Comment by robcasloz** [comment]
> If I understand it correctly, much of the diff is to ensure that `ArrayCopyNode::make` (in `BarrierSetC2::clone`) gets the correct value for the `length` arg, calculated as `align_up(array-length * elem-size, word-size) / word-size`.
> 
> I wonder if it's possible to pass the actual array length (#slots) as `length` and move the merge-bytes-to-words-copying optimization to a lower level, e.g. inside `conjoint_jbytes`. Ofc, `BarrierSetC2::clone_at_expansion` and its derived siblings need to be adjusted accordingly, e.g. to use the actual elem-type.
> 
> (Preexisting: having `ArrayCopyNode` to cover both array and instance cloning hinders the readability, IMO.)

Thanks for looking at this, Albert! I agree that the code could benefit from some clean-up, and postponing the merge-bytes-to-words-copying optimization to at least BarrierSetC2::clone_at_expansion() is worth exploring. However, your suggested refactoring would not be trivial, so I suggest to integrate this fix as-is and address the simplification in a separate RFE. Please let me know if you agree, and, if so, I will create a separate RFE for your suggestion.

**Review by albertnetymk — APPROVED** [review]
> I suggest to integrate this fix as-is and address the simplification in a separate RFE

Sounds reasonable.

**Comment by robcasloz** [comment]
> Sounds reasonable.

Thanks, Albert! I created [JDK-8314994](https://bugs.openjdk.org/browse/JDK-8314994) to capture your suggestions, please feel free to edit/extend the description if needed to reflect better your idea.

**Pull request closed without merging**